File: kt.py
import json
import sys

class TourFinder(object):
    moves = [(-2,-1), (-2,1), (-1,-2), (-1,2), (1,-2), (1,2), (2,-1), (2,1)]

    def __init__(self, n, knights_len, k, induced):
        self.n = n
        self.knights_len = knights_len
        self.k = k
        self.induced = induced
        self.square_used = [[0] * n for _ in range(n)]

    def valid_position(self, knight, value):
        if self.induced:
            for move in TourFinder.moves:
                s0 = knight[0] - move[0]
                s1 = knight[1] - move[1]
                if s0 < 0 or s0 >= n or s1 < 0 or s1 >= n:
                    continue
                square = self.square_used[s0][s1]
                if not square:
                    continue
                if square == (value[0], value[1]-1):
                    continue
                if square == (value[0]-1, value[1]):
                    continue
                return False
        return (
            knight[0] >= 0 and
            knight[0] < n and
            knight[1] >= 0 and
            knight[1] < n and
            not self.square_used[knight[0]][knight[1]]
        )

    def revert_knights(self, knights, move):
        for knight in knights:
            self.square_used[knight[0]][knight[1]] = 0
            knight[0] -= move[0]
            knight[1] -= move[1]

    def update_knights(self, knights, move, value):
        for i, knight in enumerate(knights):
            if not self.valid_position([knight[0] + move[0], knight[1] + move[1]], (value, i)):
                self.revert_knights(knights[:i], move)
                return False
            knight[0] += move[0]
            knight[1] += move[1]
            self.set_square(knight, (value, i))
        return True

    def set_square(self, knight, value):
        self.square_used[knight[0]][knight[1]] = value

    def search(self, knights, tour_len):
        if tour_len == k:
            return True
        for move in TourFinder.moves:
            if self.update_knights(knights, move, tour_len + 1):
                if self.search(knights, tour_len + 1):
                    return True
                self.revert_knights(knights, move)
        return False

    def initial_knights(self, knights_len, knights):
        if len(knights) == 0:
            for i in range(n):
                for j in range(n):
                    self.set_square([i, j], (1, 0))
                    if knights_len == 1:
                        yield [[i, j]]
                    else:
                        yield from self.initial_knights(knights_len, [[i, j]])
                    self.square_used[i][j] = 0
        else:
            for move in TourFinder.moves if len(knights) > 1 else [(1, 2)]:
                prev_knight = knights[-1]
                new_knight = [prev_knight[0] + move[0], prev_knight[1] + move[1]]
                if self.valid_position(new_knight, (1, len(knights))):
                    self.set_square(new_knight, (1, len(knights)))
                    knights.append(new_knight)
                    if knights_len == len(knights):
                        yield knights
                    else:
                        yield from self.initial_knights(knights_len, knights)
                    self.square_used[new_knight[0]][new_knight[1]] = 0
                    del knights[-1]

    def find_tour(self):
        for knights in self.initial_knights(knights_len, []):
            if self.search(knights, 1):
                return self.square_used

def format_result(n, knights_len, k, induced, result):
    mat = [[None] * knights_len for _ in range(k)]
    for i in range(n):
        for j in range(n):
            d = result[i][j]
            if d != 0:
                mat[d[0]-1][d[1]] = (i, j)
    return {"n": n, "knights": knights_len, "k": k, "induced": induced, "tour": mat}

if __name__ == "__main__":
    n = int(sys.argv[1])
    knights_len = int(sys.argv[2])
    k = int(sys.argv[3])
    induced = len(sys.argv) > 4 and sys.argv[4] == "--induced"

    if k > 0:
        result = TourFinder(n, knights_len, k, induced).find_tour()
    else:
        # optimise
        result = None
        k = 1
        while True:
            result_ = TourFinder(n, knights_len, k, induced).find_tour()
            if result_ is None:
                k -= 1
                break
            else:
                result = result_
                k += 1
    if result:
        print(json.dumps(format_result(n, knights_len, k, induced, result)))
    else:
        print("No solution exists")

File: kt.c
//
// This program is a fairly direct translation of kt.py
//

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define max_n 64
#define min_k 2
#define max_k 4

long search_calls;

int n, k, induced;  // command-line arguments

struct Pair
{
    int a;
    int b;
};

bool square_used[max_n][max_n];
struct Pair trail[max_n * max_n];
int trail_len = 0;
int incumbent_len = 0;
struct Pair knights[max_k];
int knights_len = 0;

struct Pair moves[] = {
    {1, 2},
    {-1, -2},
    {-2, -1},
    {-2, 1},
    {-1, 2},
    {1, -2},
    {2, -1},
    {2, 1}
};
int moves_len = 8;

struct Pair sum(struct Pair p, struct Pair q)
{
    return (struct Pair) {p.a + q.a, p.b + q.b};
};

struct Pair difference(struct Pair p, struct Pair q)
{
    return (struct Pair) {p.a - q.a, p.b - q.b};
};

bool equals(struct Pair p, struct Pair q)
{
    return p.a == q.a && p.b == q.b;
};

void debug()
{
    printf("------\n");
    for (int i=0; i<trail_len; i++) printf(" %d,%d", trail[i].a, trail[i].b);
    printf("\n");
    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            printf("%d ", square_used[i][j]);
        }
        printf("\n");
    }
    printf("------\n");
    printf("\n");
}

bool is_valid_square(struct Pair knight)
{
    if (knight.a < 0 || knight.a >= n || knight.b < 0 || knight.b >= n) return false;
    if (square_used[knight.a][knight.b]) return false;
    if (induced) {
        for (int i=0; i<moves_len; i++) {
            struct Pair move = moves[i];
            struct Pair s = difference(knight, move);
            if (s.a < 0 || s.a >= n || s.b < 0 || s.b >= n) continue;
            if (!square_used[s.a][s.b]) continue;
            if (trail_len > 0 && equals(s, trail[trail_len - 1])) continue;
            if (trail_len >= k && equals(s, trail[trail_len - k])) continue;
            return false;
        }
    }
    return true;
};

void set_square(struct Pair knight)
{
    square_used[knight.a][knight.b] = true;
    trail[trail_len++] = knight;
}

void unset_square()
{
    struct Pair knight = trail[--trail_len];
    square_used[knight.a][knight.b] = false;
}

void revert_knights(int count, struct Pair move)
{
    for (int i=0; i<count; i++) {
        unset_square();
        knights[i] = difference(knights[i], move);
    }
}

bool update_knights(struct Pair move)
{
    for (int i=0; i<k; i++) {
        struct Pair knight = knights[i];
        if (!is_valid_square(sum(knight, move))) {
            revert_knights(i, move);
            return false;
        }
        knights[i] = sum(knight, move);
        set_square(knights[i]);
    }
    return true;
}

void search(int tour_len)
{
    ++search_calls;
    if (tour_len > incumbent_len) {
        incumbent_len = tour_len;
        for (int i=0; i<trail_len; i++) printf(" %d,%d", trail[i].a, trail[i].b);
        printf("\n");
    }
    for (int i=0; i<moves_len; i++) {
        struct Pair move = moves[i];
        if (update_knights(move)) {
            search(tour_len + 1);
            revert_knights(k, move);
        }
    }
}

void place_m_knights_and_search()
{
    if (knights_len == k) {
        search(1);
    } else {
        for (int i=0; i<moves_len; i++) {
            if (knights_len == 1 && i != 0) {
                // Symmetry breaking; use move {1,2}
                continue;
            }
            struct Pair move = moves[i];
            struct Pair prev_knight = knights[knights_len - 1];
            struct Pair new_knight = sum(prev_knight, move);
            if (is_valid_square(new_knight)) {
                set_square(new_knight);
                knights[knights_len++] = new_knight;
                place_m_knights_and_search();
                --knights_len;
                unset_square();
            }
        }
    }
}

void find_maximum_tour()
{
    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            knights[knights_len++] = (struct Pair) {i, j};
            set_square((struct Pair) {i, j});
            place_m_knights_and_search();
            unset_square();
            --knights_len;
        }
    }
    printf("Max m is %d\n", incumbent_len);
}

int main(int argc, char *argv[])
{
    if (argc != 4 || sscanf(argv[1], "%d", &n) != 1
            || sscanf(argv[2], "%d", &k) != 1 || sscanf(argv[3], "%d", &induced) != 1) {
        fprintf(stderr, "Usage: %s n k induced    (where induced is 0 or 1)\n", argv[0]);
        exit(-1);
    }
    if (n < 3 || n > max_n) {
        fprintf(stderr, "n must be between 3 and %d!\n", max_n);
        exit(-2);
    }
    if (k < min_k || k > max_k) {
        fprintf(stderr, "k must be between %d and %d!\n", min_k, max_k);
        exit(-3);
    }
    printf("%dxm knights in a %dx%d square\n", k, n, n);
    find_maximum_tour();
    printf("%ld search() calls\n", search_calls);
}
